File: frontend/src/model/Questions.ts
export interface Questions {
    questions: Question[]
}

export type Question = ChoiceQuestion | OpenQuestion


export class ChoiceQuestion {
    questionUUID: string
    questionTitle: string
    questionDescription: string | null
    points: number | null
    possibleAnswersToIfCorrect: PossibleAnswersToIfCorrect[]

    picture: string | null

    constructor(questionUUID: string,
                questionTitle: string,
                questionDescription: string | null,
                points: number | null,
                possibleAnswersToIfCorrect: PossibleAnswersToIfCorrect[],
                picture: string | null
    ) {
        this.questionUUID = questionUUID
        this.questionTitle = questionTitle
        this.questionDescription = questionDescription
        this.points = points
        this.possibleAnswersToIfCorrect = possibleAnswersToIfCorrect
        this.picture = picture
    }
}

export interface PossibleAnswersToIfCorrect {
    text: string
    uuid: string
    correct: boolean | false
}

export class OpenQuestion {
    questionUUID: string
    questionTitle: string
    questionDescription: string | null
    points: number | null
    picture: string | null

    constructor(questionUUID: string,
                questionTitle: string,
                questionDescription: string | null,
                points: number | null,
                picture: string | null
    ) {
        this.questionUUID = questionUUID
        this.questionTitle = questionTitle
        this.questionDescription = questionDescription
        this.points = points
        this.picture = picture
    }
}

export enum QuestionType {
    OPEN_QUESTION = "OPEN_QUESTION",
    CHOICE_QUESTION = "CHOICE_QUESTION"
}

export const translationOfQuestionType = {
    OPEN_QUESTION: 'Въпрос с отворен отговор',
    CHOICE_QUESTION: "Въпрос със затворен отговор"
}
export const isChoiceQuestion = (question: Question): question is ChoiceQuestion =>
        question instanceof ChoiceQuestion || (question && 'possibleAnswersToIfCorrect' in question)

export const isOpenQuestion = (question: Question): question is OpenQuestion =>
        question instanceof OpenQuestion || (question && !('possibleAnswersToIfCorrect' in question))

export const defineQuestionType = (question: Question) => {
    if (isOpenQuestion(question)) {
        return QuestionType.OPEN_QUESTION
    } else if (isChoiceQuestion(question)) {
        return QuestionType.CHOICE_QUESTION
    } else {
        return null
    }
}

File: backend/src/main/kotlin/com/nevexis/backend/schoolManagement/assignments/examination/Questions.kt
package com.nevexis.backend.schoolManagement.assignments.examination

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject

@Serializable
data class Questions(
    val questions: List<Question>
)

@Serializable(with = QuestionSerializer::class)
sealed class Question {

    abstract val questionUUID: String
    abstract val questionTitle: String
    abstract val questionDescription: String?
    abstract val points: Int

    @Serializable
    data class ChoiceQuestion(
        override val questionUUID: String,
        val possibleAnswersToIfCorrect: List<PossibleAnswersToIfCorrect>,
        override val questionTitle: String,
        override val questionDescription: String?,
        override val points: Int
    ) : Question()

    @Serializable
    data class OpenQuestion(
        override val questionUUID: String,
        override val questionTitle: String,
        override val questionDescription: String?,
        override val points: Int
    ) : Question()

}

@Serializable
data class PossibleAnswersToIfCorrect(
    val text: String,
    val uuid: String,
    val correct: Boolean = false
)

object QuestionSerializer : JsonContentPolymorphicSerializer<Question>(Question::class) {
    override fun selectDeserializer(element: JsonElement) = when {
        "possibleAnswersToIfCorrect" in element.jsonObject -> {
            println("Detected ChoiceQuestion")
            Question.ChoiceQuestion.serializer()
        }

        else -> {
            println("Detected OpenQuestion")
            Question.OpenQuestion.serializer()
        }
    }
}

//class QuestionDeserializer : JsonDeserializer<Question>() {
//    override fun deserialize(
//        p: com.fasterxml.jackson.core.JsonParser,
//        ctxt: DeserializationContext
//    ): Question {
//        val rootNode = p.codec.readTree<JsonNode>(p) ?: throw IllegalArgumentException("Unable to parse JSON")
//        val evaluationNode = rootNode.path("question")
//
//        if (evaluationNode.isMissingNode) {
//            throw IllegalArgumentException("Assignment value is missing in the JSON")
//        }
//
//        val questionJson = rootNode.toString()
//
//        return when {
//            questionJson.contains("possibleAnswersToIfCorrect") -> Json.decodeFromString<Question.ChoiceQuestion>(
//                questionJson
//            )
//
//            else -> Json.decodeFromString<Question.OpenQuestion>(questionJson)
//        }
//    }
//}